### PR TITLE
SPOI-11620 Batch Support for File Input operator

### PR DESCRIPTION
**Pull request by ajaygit158**

### CODE REVIEW

**Comment by ajaygit158** [comment]
@bhupeshchawda Kindly review.

File: operators/library/src/main/java/com/datatorrent/batch/EndApplicationControlTuple.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+* BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF
+* OR IN CONNECTION WITH THE LICENSED MATERIALS.
+ */
+package com.datatorrent.batch;
+
+import org.apache.apex.api.operator.ControlTuple;
+import org.apache.hadoop.classification.InterfaceStability.Evolving;
+
+@Evolving
+public interface EndApplicationControlTuple extends ControlTuple
```
Comment (bhupeshchawda) [inline operators/library/src/main/java/com/datatorrent/batch/EndApplicationControlTuple.java]: Is End Application control needed?

File: operators/library/src/main/java/com/datatorrent/io/fs/AbstractBatchFileInputOperator.java [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+  public abstract void emitEndApplicationControlTuple();
+
+  /**
+   * The class that is used to scan for new files in the directory for the
+   * AbstractBatchFileInputOperator. By default, if batch is enabled, it will
+   * perform a single scan of the input directory and emit a single batch only
+   */
+  @SuppressWarnings("serial")
+  @Evolving
+  public static class BatchDirectoryScanner extends com.datatorrent.lib.io.fs.AbstractFileInputOperator.DirectoryScanner
```
Comment (bhupeshchawda) [inline operators/library/src/main/java/com/datatorrent/io/fs/AbstractBatchFileInputOperator.java]: I think we should design a generic scanner interface, an implementation of which would decide how the scan would happen. For example we may have a SingleScanDirectoryScanner or a TimeBasedDirectoryScanner etc.
The scanner can then be plugged instead of the initial scanner in the AbstractFileInputOperator class.
The respective implementation would then scan depending on their behavior. We can add another method in the scanner - something like isDone() which would tell the operator whether to end the batch or continue..

File: operators/library/src/main/java/com/datatorrent/io/fs/AbstractBatchFileInputOperator.java [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+    {
+      LinkedHashSet<Path> fileList = super.scan(fs, filePath, consumedFiles);
+      scanCount++;
+      if (shouldEndScan()) {
+        allFilesScanned = true;
+      }
+      return fileList;
+    }
+
+    public enum ScannerState
```
Comment (bhupeshchawda) [inline operators/library/src/main/java/com/datatorrent/io/fs/AbstractBatchFileInputOperator.java]: This seems to be too much bookkeeping, I would suggest to remove this..

File: operators/library/src/main/java/com/datatorrent/io/fs/AbstractBatchFileInputOperator.java [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+  /**
+   * By setting this property to true, a single scan will be performed every
+   * batchIntervalMillis. All files made available in this scan period will form
+   * a part of same batch. For each batch, start batch and end batch control
+   * tuples will be emitted before and after the batch respectively. Note that
+   * batch tuple processing currently works only in single operator instance
+   * scenarios i.e without partitions. However, batch processing with parallel
+   * partitions will work
+   */
+  private boolean emitBatchTuples = false;
```
Comment (bhupeshchawda) [inline operators/library/src/main/java/com/datatorrent/io/fs/AbstractBatchFileInputOperator.java]: we don't need this right? we'll anyway emit them at start and end..

File: operators/library/src/main/java/com/datatorrent/io/fs/AbstractBatchFileInputOperator.java [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+      startBatchEmitted = false;
+      endBatchEmitted = false;
+      endApplicationEmitted = false;
+      scannerState = null;
+    }
+  }
+
+  public AbstractBatchFileInputOperator()
+  {
+    BatchDirectoryScanner batchScanner = new BatchDirectoryScanner();
```
Comment (bhupeshchawda) [inline operators/library/src/main/java/com/datatorrent/io/fs/AbstractBatchFileInputOperator.java]: this should be a parameter

File: operators/library/src/main/java/com/datatorrent/io/fs/AbstractBatchFileInputOperator.java [hunk 6]
```diff
@@ -0,0 +1,454 @@
...
+    this.scanner = batchScanner;
+  }
+
+  @Override
+  public void setup(OperatorContext context)
+  {
+    super.setup(context);
+    if (scanner instanceof BatchDirectoryScanner) {
```
Comment (bhupeshchawda) [inline operators/library/src/main/java/com/datatorrent/io/fs/AbstractBatchFileInputOperator.java]: we should not do anything special here..batch is default...just end depending on when the scanner tells you..the streaming scanner would never tell you to end..

File: operators/library/src/main/java/com/datatorrent/batch/BatchControlTuple.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+* OR IN CONNECTION WITH THE LICENSED MATERIALS.
+ */
+
+package com.datatorrent.batch;
+
+import org.apache.apex.api.operator.ControlTuple;
+import org.apache.hadoop.classification.InterfaceStability.Evolving;
+
+@Evolving
+public interface BatchControlTuple extends ControlTuple
```
Comment (ajaygit158) [inline operators/library/src/main/java/com/datatorrent/batch/BatchControlTuple.java]: This doesn't work if public is removed (Compile time issues).

File: operators/library/src/main/java/com/datatorrent/batch/EndApplicationControlTuple.java (hunk 1 above)
Comment (ajaygit158) [inline operators/library/src/main/java/com/datatorrent/batch/EndApplicationControlTuple.java]: Removed for now. Directly calling shutdown.

File: operators/library/src/main/java/com/datatorrent/io/fs/AbstractBatchFileInputOperator.java (hunk 4 above)
Comment (ajaygit158) [inline operators/library/src/main/java/com/datatorrent/io/fs/AbstractBatchFileInputOperator.java]: I had added this so that this operator can be used for both streaming and batch usecases. Removing it for now.

File: operators/library/src/main/java/com/datatorrent/io/fs/AbstractBatchFileInputOperator.java (hunk 5 above)
Comment (ajaygit158) [inline operators/library/src/main/java/com/datatorrent/io/fs/AbstractBatchFileInputOperator.java]: The default scanner gets set to the DirectoryScanner present in parent class if this piece of code is not added. We have provided setters and getters so that the scanner can be changed. Let me know if there is a way around this.

**Comment by ajaygit158** [comment]
@bhupeshchawda Made the changes suggested by you. Kindly review!

**Comment by ajaygit158** [comment]
@bhupeshchawda Incorporated the changes suggested.

File: pom.xml [hunk 8]
```diff
@@ -125,9 +125,9 @@
   </dependencies>
   
   <modules>
-  	<module>app-templates</module>
-  	<module>solutions</module>
-  	<module>operators</module>
```
Comment (bhupeshchawda) [inline pom.xml]: why this change?
